compiler/optimizer/VPHandlers.cpp

```diff
@@ -5239,6 +5238,98 @@ TR::Node *constrainCall(TR::ValuePropagation *vp, TR::Node *node)
       bool isFloatAbs =   (symbol->getRecognizedMethod() == TR::java_lang_Math_abs_F);
       bool isDoubleAbs =  (symbol->getRecognizedMethod() == TR::java_lang_Math_abs_D);
 
+      bool isGetModifiersImpl = (symbol->getRecognizedMethod() == TR::java_lang_Class_getModifiersImpl);
+      bool isIsAssignableFrom = (symbol->getRecognizedMethod() == TR::java_lang_Class_isAssignableFrom);
+      bool isIsInterface = (symbol->getRecognizedMethod() == TR::java_lang_Class_isInterface);
+      bool isIdentityHashCode = (symbol->getRecognizedMethod() == TR::java_lang_J9VMInternals_identityHashCode);
+
+      if (isGetModifiersImpl)
+         {
+         TR::Node *classChild = node->getFirstChild();
+         bool classChildGlobal;
+         TR::VPConstraint *classChildConstraint = vp->getConstraint(classChild, classChildGlobal);
+         if (classChildConstraint && classChildConstraint->isJavaLangClassObject() == TR_yes
+             && classChildConstraint->isNonNullObject()
+             && classChildConstraint->getClassType()
+             && classChildConstraint->getClassType()->asFixedClass())
+            {
+            int32_t modifiersForClass = vp->comp()->fej9vm()->getRawModifiers(classChildConstraint->getClass());
+            TR::VPConstraint *modifiersConstraint = TR::VPIntConst::create(vp, modifiersForClass);
+            vp->replaceByConstant(node, modifiersConstraint, classChildGlobal);
+            TR::DebugCounter::incStaticDebugCounter(vp->comp(), TR::DebugCounter::debugCounterName(vp->comp(), "constrainCall/%s", "getModifiersImpl"));
+            return node;
+            }
+         }
+      else if (isIsInterface)
+         {
+         TR::Node *classChild = node->getLastChild();
+         bool classChildGlobal;
+         TR::VPConstraint *classChildConstraint = vp->getConstraint(classChild, classChildGlobal);
+         if (classChildConstraint && classChildConstraint->isJavaLangClassObject() == TR_yes
+             && classChildConstraint->isNonNullObject()
+             && classChildConstraint->getClassType()
+             && classChildConstraint->getClassType()->asFixedClass())
+            {
+            int32_t isInterface = TR::Compiler->cls.isInterfaceClass(vp->comp(), classChildConstraint->getClass());
+            TR::VPConstraint *isInterfaceConstraint = TR::VPIntConst::create(vp, isInterface);
+            vp->replaceByConstant(node, isInterfaceConstraint, classChildGlobal);
+            TR::DebugCounter::incStaticDebugCounter(vp->comp(), TR::DebugCounter::debugCounterName(vp->comp(), "constrainCall/%s", "isInterface"));
+            return node;
+            }
+         }
+      else if (isIsAssignableFrom)
```

Comment:
You can probably get more milage here - we don't need both classes to be fixed classes - it may be sufficient to have a fixed class and a class constraint holding a top type...